src/tests/system/tests/test_gpo.py

```diff
@@ -0,0 +1,1056 @@
+        2. Create and link the GPO 'site policy' and add 'Domain Admins' to SeInteractiveLogonRight key
+        3. Configure sssd.conf with 'ad_gpo_access_control' = 'enforcing'
+        4. Start SSSD
+    :steps:
+        1. Authenticate 'user'
+        2. Check logs for gpo_access_check result
+    :expectedresults:
+        1. User authentication is unsuccessful
+        2. Access check result is not denied or granted
+    :customerscenario: True
+    """
+    ad.user("user").add()
+
+    ad.gpo("site policy").add().policy(
+        {
+            "SeInteractiveLogonRight": [ad.group("Domain Admins")],
+            "SeDenyInteractiveLogonRight": [],
+        }
+    ).link()
+
+    client.sssd.domain["ad_gpo_access_control"] = "enforcing"
+    client.sssd.start()
+
+    assert not client.auth.parametrize(method).password(username="user", password="Secret123")
+
+    log_str = client.fs.read(client.sssd.logs.domain())
+    assert "access_denied = 0" in log_str
+    assert "access_granted = 0" in log_str
+
+
+@pytest.mark.importance("critical")
+@pytest.mark.parametrize("method", ["su", "ssh"])
+@pytest.mark.topology(KnownTopology.AD)
+def test_gpo__is_set_to_enforcing_with_no_policy(client: Client, ad: AD, method: str):
+    """
+    :title: Group policy object host base access control is set to enforcing with no policy
+    :description:
+        When ad_gpo_access_control is set to enforcing with no policy applied. It should
+        permit all users to login.
+    :setup:
+        1. Create the following user, 'user'
+        2. Configure sssd.conf with 'ad_gpo_access_control' = 'enforcing'
+        3. Start SSSD
+    :steps:
+        1. Authenticate 'user'
+        2. Check logs for gpo_access_check result
+    :expectedresults:
+        1. User authentication is successful
+        2. Access check result is granted
+    :customerscenario: True
+    """
+    ad.user("user").add()
+
+    client.sssd.domain["ad_gpo_access_control"] = "enforcing"
+    client.sssd.start()
+
+    assert client.auth.parametrize(method).password(username="user", password="Secret123")
+
+    log_str = client.fs.read(client.sssd.logs.domain())
+    assert "access_granted = 1" in log_str
+
+
+@pytest.mark.importance("critical")
+@pytest.mark.parametrize("method", ["su", "ssh"])
+@pytest.mark.topology(KnownTopology.AD)
+def test_gpo__is_set_to_permissive_and_users_are_allowed(client: Client, ad: AD, method: str):
+    """
+    :title: Group policy object host base access control is set to permissive
+    :description:
+        When ad_gpo_access_control is set to permissive, all users are able to login despite what
+        is configured in the policy. The rules are still processed and can be viewed in the logs.
+    :setup:
+        1. Create the following user, 'user1'
+        2. Create and link the GPO 'site policy' and add 'user1', group and 'Domain Admins' to
+           SeInteractiveLogonRight key. Add 'deny_user1' to SeDenyInteractiveLogonRight key
```

Comment:
Looks like you wanted to add a user which would be denied if GPOs are enforced. This is imo a good idea, but it is missing in the test.

Reply:
Ah, you are doing this in the next test. I think this split of tests is too much. Imo ideally all tests related to access control have some successful and unsuccessful logins to make sure there is a real access control.

The `permissive` tests are special, since all users are allowed, however having rules  which would allow some and deny others and show that users from both groups are allowed in a single tests would be best here.

Reply:
So it's funny you say that the team prefers the broken-up, specific test cases. There were enough people who opposed my suggestions to have fewer test cases but more done in each test that I didn't fight it. 

Below, you can see the original test before this conversation happened; it's much closer to your suggestion. So I had to laugh. 

```
def test_gpo__is_set_to_permissive(client: Client, ad: AD, sssd_service_user: str):
    """
    :title: Group policy object host base access control is set to permissive
    :description:
        When ad_gpo_access_control is set to permissive, all users are able to login despite what
        is configured in the policy. The rules are still processed and can be viewed in the logs.
    :setup:
        1. Create the following users; 'user', 'user1', 'deny_user1'
        2. Create and link the GPO 'test policy' and add 'user1', group and 'Domain Admins' to
           SeInteractiveLogonRight key. Add 'deny_user1' to SeDenyInteractiveLogonRight key
        4. Configure sssd.conf with 'ad_gpo_access_control' = 'permissive'
        5. Start SSSD
    :steps:
        1. Authenticate 'user', 'user1' and 'deny_user1' with both su and ssh
        2. Check logs to ensure that SSSD is in permissive
    :expectedresults:
        1. All user authentication is successful
        2. SSSD is in permissive
    :customerscenario: True
    """
    ad.user("user").add()
    user1 = ad.user("user1").add()
    deny_user1 = ad.user("deny_user1").add()

    ad.gpo("test policy").add().policy(
        {
            "SeInteractiveLogonRight": [user1, ad.group("Domain Admins")],
            "SeDenyInteractiveLogonRight": [deny_user1],
        }
    ).link()

    client.sssd.set_service_user(sssd_service_user)
    client.sssd.domain["ad_gpo_access_control"] = "permissive"
    client.sssd.start()

    assert client.auth.ssh.password(username="user1", password="Secret123")
    assert client.auth.su.password(username="user1", password="Secret123")
    assert client.auth.ssh.password(username="user", password="Secret123")
    assert client.auth.su.password(username="user", password="Secret123")
    assert client.auth.ssh.password(username="deny_user1", password="Secret123")
    assert client.auth.su.password(username="deny_user1", password="Secret123")

    log_str = client.fs.read(client.sssd.logs.domain())
    assert "Option ad_gpo_access_control has value permissive" in log_str, f"Log file has wrong format: {log_str}"
```

Reply:
I'm sorry, I thought I added the access evaluations to permissive, to check that 'user1' is granted access, but they're missing from all the permissive tests. I don't mind merging all the tests, but if they need to be evaluated, I think the team will want to keep them separated. 

Do we need to assert that 'user1' is 'access_granted =1' is important to permissive? I'm kind of indifferent so I'll do whatever you think is best.

Reply:
I'm sorry, I thought I added the access evaluations to permissive, to check that 'user1' is granted access, but they're missing from all the permissive tests. I don't mind merging all the tests, but if they need to be evaluated, I think the team will want to keep them separated. 

Do we need to assert that 'user1' is 'access_granted =1' is important to permissive? I'm kind of indifferent so I'll do whatever you think is best.

Reply:
Hi,

I have to admit that I didn't look into the length at the individual tests when saying that I prefer small tests. However, as mentioned in other comments, I think a single access control test should contain evidence that expected objects are allowed  and other expected are rejected with a given configuration. @jakub-vavra-cz , iirc you were also in favor of having small tests, what's your opinion?


But I think you can split the test above differently. `ssh` and `su` are handled by two different maps/policies, `ad_gpo_map_remote_interactive` and `ad_gpo_map_interactive` respectively. Since you are only setting the `interactive` one this test tests login in `permissive` mode with an explicit rule given (`su`) and login in `permissive` mode without an explicit rule given (`ssh`).

bye,
Sumit

Reply:
Ah, you are using `ssh` and `su` as parameters in the current tests. Then is might be good to have a comment in the description about the difference?

Reply:
Also, FYI, in the test, if SeRemoteInteractive is omitted, it copies the value from SeInteractive, so by default it will be set for both mappings. I don't think having the other rule there, hurts the test at all.

Reply:
Added the comments to the file.

Reply:
> Also, FYI, in the test, if SeRemoteInteractive is omitted, it copies the value from SeInteractive, so by default it will be set for both mappings. I don't think having the other rule there, hurts the test at all.

Hi,

this sounds like a good feature of the framework when e.g. testing authentication with different services to make sure the user is granted access with minimal effort in the test. But when testing the GPO based access control directly I find this magic behind the scenes a bit surprising.

Would it be possible to add new methods so that other tests can easily add users to multiple GPOs/maps but also have methods for the GPO tests where everything has to be set explicitly?

bye,
Sumit